lib/Sequence.fz

```diff
@@ -903,9 +903,18 @@ public Sequence(public T type) ref : property.equatable is
   # equality
   #
   public fixed redef type.equality(a, b Sequence T) =>
+    equality0(a1, b1 Sequence (T2 : property.equatable)) =>
+      ((a1.zip b1 x,y->x=y) ∀ x->x)
+
     if T : property.equatable
-      a.count = b.count
-        && ((a.zip b x,y->x=y) ∀ x->x)
+      if a.finite && b.finite
+        a.count = b.count && equality0 a b
+      else if a.finite
+        a.count = (b.take (a.count+1)).count && equality0 a b
+      else if b.finite
+        b.count = (a.take (b.count+1)).count && equality0 a b
+      else
+        equality0 a b
```

Comment:
I think this code is overly complex and not needed. It is suffcient to just use `equality0`:
```suggestion
      equality0 a b
```
or, if we want to optimize the case when both are known to be finite but might differ in their count: 
```suggestion
      !(a.finite && b.finite && a.count != b.count) && equality0 a b
```

Reply:
you are right, the code is needed.